dpos/state/state.go

```diff
@@ -802,98 +831,86 @@ func (s *State) ProcessSpecialTxPayload(p types.Payload) {
 	s.history.commit(0)
 }
 
-// tryLeaveInactiveMode for inactive arbitrators to try leaving inactive mode if
-// over the inactive duration
-func (s *State) tryLeaveInactiveMode(blockHeight uint32) {
-	tryLeaveInactive := func(key string, producer *Producer) {
-		s.history.append(blockHeight, func() {
-			if blockHeight > producer.inactiveSince+s.chainParams.InactiveDuration {
-				pk, err := common.HexStringToBytes(key)
-				if err != nil {
-					log.Warn("[tryLeaveInactiveMode] convert to public key" +
-						" byte error")
-					return
-				}
-
-				producer.state = Activate
+// setInactiveProducer set active producer to inactive state
+func (s *State) setInactiveProducer(producer *Producer, key string,
+	height uint32) {
+	producer.inactiveSince = height
+	producer.state = Inactivate
+	s.inactiveProducers[key] = producer
+	delete(s.activityProducers, key)
 
-				if !s.arbiters.IsCRCArbitrator(pk) {
-					s.activityProducers[key] = producer
-				}
-				delete(s.inactiveProducers, key)
-			}
-		}, func() {
-			if blockHeight <= producer.inactiveSince+s.chainParams.InactiveDuration {
+	producer.penalty += s.chainParams.InactivePenalty
+}
 
-				producer.state = Inactivate
-				s.inactiveProducers[key] = producer
-				delete(s.activityProducers, key)
-			}
-		})
-	}
+// revertSettingInactiveProducer revert operation about setInactiveProducer
+func (s *State) revertSettingInactiveProducer(producer *Producer, key string,
+	height uint32) {
+	producer.inactiveSince = 0
+	producer.state = Activate
+	s.activityProducers[key] = producer
+	delete(s.inactiveProducers, key)
 
-	for k, v := range s.inactiveProducers {
-		tryLeaveInactive(k, v)
+	if producer.penalty < s.chainParams.InactivePenalty {
```

Comment:
No if else in history func

Reply:
it's just a guard condition to prevent the penalty from becoming negative